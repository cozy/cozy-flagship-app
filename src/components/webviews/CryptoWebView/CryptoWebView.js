import React, { useEffect, useRef } from 'react'
import { View } from 'react-native'
import { WebView } from 'react-native-webview'

import {
  subscribeToCrypto,
  unsubscribeFromCrypto,
  sendAnswer
} from './cryptoObservable/cryptoObservable'
import { html } from './jsInteractions/jsCryptoInjection'

/**
 * Lorsque la webview se rend, elle initialise un script JS de hash de mot de passe
 * Ce script est appelé par un observer "CryptoObservable"
 * La webview s'enregistre auprès du CryptoObservable par `subscribeToCrypto` et lui fournit un callback
 * Lorsque ce callback est appelé par l'observable, la crypto transfert le message au JS qui calcule le hash
 * et le retourne par postMessage
 * Dans le postMessage on appelle sendAnswer() avec le mot de passe hashé
 */
export const CryptoWebView = () => {
  const webviewRef = useRef()

  const processMessage = (message, messageId, param) => {
    const payload = JSON.stringify({
      message,
      messageId,
      param
    })

    const webView = webviewRef.current
    webView.postMessage(payload)
  }

  const processAnswer = event => {
    const webviewAnswer = JSON.parse(event.nativeEvent.data)
    sendAnswer(webviewAnswer)
  }

  useEffect(() => {
    subscribeToCrypto(processMessage)

    return () => {
      unsubscribeFromCrypto(processMessage)
    }
  }, [webviewRef])

  return (
    <View style={{ height: 0 }}>
      <WebView
        ref={webviewRef}
        javaScriptEnabled={true}
        onMessage={processAnswer}
        originWhitelist={['*']}
        source={{ html, baseUrl: 'https://localhost' }}
      />
    </View>
  )
}
