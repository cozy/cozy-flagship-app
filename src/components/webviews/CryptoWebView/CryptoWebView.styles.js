export const styles = {
  cryptoView: {
    height: 0
  }
}
