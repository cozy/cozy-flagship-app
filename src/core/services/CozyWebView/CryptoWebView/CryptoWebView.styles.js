import { StyleSheet } from 'react-native'

export const styles = StyleSheet.create({
  cryptoView: {
    height: 0
  }
})
