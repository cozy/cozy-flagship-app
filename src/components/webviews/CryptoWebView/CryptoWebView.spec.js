import { CryptoWebView } from './CryptoWebView'
import React from 'react'
import { render } from '@testing-library/react-native'
import {
  sendAnswer,
  subscribeToCrypto,
  unsubscribeFromCrypto
} from './cryptoObservable/cryptoObservable'

jest.mock('./cryptoObservable/cryptoObservable', () => ({
  sendAnswer: jest.fn(),
  subscribeToCrypto: jest.fn(),
  unsubscribeFromCrypto: jest.fn()
}))

jest.mock('react-native-webview', () => {
  const React = require('react')
  class WebView extends React.Component {
    postMessage(payload) {
      this.props.onMessage({
        nativeEvent: {
          data: JSON.stringify({
            answer: 'SOME_ANSWER',
            payloadRefForTests: JSON.parse(payload)
          })
        }
      })
    }
    render() {
      return <div>WebView</div>
    }
  }
  return { WebView }
})

describe('CryptoWebview', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('should render', () => {
    // When
    const { toJSON } = render(<CryptoWebView />)
    // Then
    expect(toJSON()).toMatchSnapshot()
  })

  it('should subscribeToCrypto on mount', () => {
    // When
    render(<CryptoWebView />)
    // Then
    expect(subscribeToCrypto).toHaveBeenCalledWith(expect.any(Function))
  })

  it('should unsubscribeFromCrypto on unmount', () => {
    // Given
    const { unmount } = render(<CryptoWebView />)

    // When
    unmount()

    // Then
    expect(unsubscribeFromCrypto).toHaveBeenCalledWith(expect.any(Function))
  })

  it('should sendAnswer with expected answer from WebView', () => {
    let cbPointer
    const message = 'MESSAGE'
    const messageId = 'MESSAGEID'
    const param = { param: 'PARAM' }
    subscribeToCrypto.mockImplementation(cb => {
      cbPointer = cb
    })

    render(<CryptoWebView />)

    cbPointer(message, messageId, param)

    expect(sendAnswer).toHaveBeenCalledWith({
      answer: 'SOME_ANSWER',
      payloadRefForTests: {
        message: 'MESSAGE',
        messageId: 'MESSAGEID',
        param: { param: 'PARAM' }
      }
    })
  })
})
