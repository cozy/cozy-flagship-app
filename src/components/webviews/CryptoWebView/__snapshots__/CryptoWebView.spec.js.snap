// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CryptoWebview should render 1`] = `
<View
  style={
    Object {
      "height": 0,
    }
  }
>
  <div>
    WebView
  </div>
</View>
`;
